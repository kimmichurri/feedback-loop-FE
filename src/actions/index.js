export const hasError = (message) => ({
  type: 'HAS_ERROR',
  message
})

export const isLoading = (boolean) => ({
  type: 'IS_LOADING',
  isLoading: boolean
})

export const setSurvey = (survey) => ({
  type: 'SET_SURVEY',
  survey
})

export const setCohorts = (cohorts) => ({
  type: 'SET_COHORTS',
  cohorts
})

export const setUser = (user) => ({
  type: 'SET_USER',
  user
})

export const setCurrentCohort = (cohort) => ({
  type: 'SET_CURRENT_COHORT',
  cohort
})