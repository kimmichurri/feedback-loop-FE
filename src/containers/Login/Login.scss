.login-landing {
  background: $white;
  width: 40%;
  margin: auto;
  padding: 25px;
  border-radius: 5px;
  border-top: .65rem solid $main-blue;
}

.login-form {
  margin: 20px 0px;
}

.login-input {
  height: 40px;
  padding-left: 10px;
  padding-right: 10px;
  font-size: 16px;
  border-radius: 5px;
  margin-top: 5px;
    &:focus {
      outline: none;
    }
}

.email,
.password,
.full-name {
  display: flex;
  flex-direction: column;
  text-align: left;
  margin-bottom: 20px;
}

.role-input {
  display: flex;
  align-items: center;
  height: 40px;
  padding-left: 10px
}

.role-choice {
  display: flex;
  margin-left: 20px;
}

.role-input-p {
  margin: 0px;
  font-size: 16px;
}

.login-button {
  background-color: $main-blue;
  border-radius: 50px;
  border: none;
  color: $accent-teal;
  font-family: $main-font;
  font-weight: bold;
  font-size: 16px;
  height: 50px;
  width: 100%;
  margin-bottom: 20px;
  &:hover {
    cursor: pointer;
    background-color: $main-blue-hover;
  }
}
