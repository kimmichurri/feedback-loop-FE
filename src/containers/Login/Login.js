import React, { Component } from 'react'
import { connect } from 'react-redux'
import { setUser, setRole } from '../../actions'
import RegisterForm from '../RegisterForm/RegisterForm'
import LoginForm from '../LoginForm/LoginForm'
import { handlePost } from '../../thunks/handlePost'
import { withRouter } from 'react-router-dom'

export class Login extends Component {
  constructor() {
    super()
    this.state = {
        newUser: false
    }
  }

  createAccount = () => {
    this.setState({
      newUser: true
    })
  }

  handleLogin = (user) => {
    if(!this.state.newUser) {
      this.loginUser(user)
    } else {
      this.registerUser(user)
    }
  }

  loginUser = async (user) => {
    const url = "https://turing-feedback-api.herokuapp.com/api/v1/users/login"
    const options = {
      method: 'POST',
      body: JSON.stringify({
        email: user.emailInput,
        password: user.passwordInput
      }),
      headers: {
        'Content-Type': 'application/json'
      }
    }
    this.props.handlePost(url, options)
    const data = await this.props.handlePost(url, options)
    this.handleUser(data)
  }

  registerUser = async (user) => {
    const url = "https://turing-feedback-api.herokuapp.com/api/v1/users/register"
    const options = {
      method: 'POST',
      body: JSON.stringify({
        fullName: user.full_name,
        role: user.role,
        email: user.email,
        password: user.password_1
      }),
      headers: {
        'Content-Type': 'application/json'
      }
    }
    const data = await this.props.handlePost(url, options)
    this.handleUser(data)
  }

  handleUser = (data) => {
    const validUser = data.api_key
    const userRole = data.role
    this.props.setUser(validUser)
    this.props.setRole(userRole)
    this.handleRedirect()
  }

  handleRedirect = () => {
    if (this.props.role === 'Student') {
      //We'll want to change this to push to StudentDashboard
      this.props.history.push('/dashboard')
    } else {
      this.props.history.push('/dashboard')
    }
  }

  render() {
    return(
      <div className='login-landing'>
        {!this.state.newUser && 
          <LoginForm 
            createAccount={this.createAccount}
            handleLogin={this.handleLogin}
          />}
        {this.state.newUser && 
          <RegisterForm 
            handleLogin={this.handleLogin}
          />}
      </div>
    )
  }
}

export const mapStateToProps = (state) => ({
  role: state.role
})

export const mapDispatchToProps = (dispatch) => ({
  setUser: (user) => dispatch(setUser(user)),
  handlePost: (url, options) => dispatch(handlePost(url, options)),
  setRole: (role) => dispatch(setRole(role))
})

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(Login))