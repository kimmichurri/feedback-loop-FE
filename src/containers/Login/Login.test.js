import React from 'react'
import { shallow } from 'enzyme'
import { Login, mapDispatchToProps } from './Login'
import { setUser, setRole } from '../../actions'
jest.mock('../../thunks/handlePost')

describe('Login', () => {
  let mockSetUser
  let mockSetRole
  let mockUser
  let wrapper
  let mockHistory = { push: jest.fn() }

  beforeEach(() => {
    mockSetUser = jest.fn()
    mockSetRole = jest.fn()
    mockUser = {
      "api_key": "6d07603f-c6d8-47e7-b48b-2fea92eb51e5",
      "full_name": "Sample User",
      "id": 2,
      "role": "Student"
    }
  

    wrapper = shallow(
      <Login 
        setUser={mockSetUser}
        setRole={mockSetRole}
        handlePost={jest.fn()}
        history={mockHistory}
      />
    )
  })

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should have a default state', () => {
    expect(wrapper.state('newUser')).toBe(false)
  })

  it('should set state when createAccount is called', () => {
    wrapper.instance().createAccount()
    expect(wrapper.state('newUser')).toBe(true)
  })

  it('should call loginUser when handleLogin is called', () => {
    const loginUserSpy = jest.spyOn(wrapper.instance(), 'loginUser')
    wrapper.setState({newUser: false})
    wrapper.instance().handleLogin()
    expect(loginUserSpy).toHaveBeenCalled()
  })

  it('should call registerUser when handleLogin is called', () => {
    const registerUserSpy = jest.spyOn(wrapper.instance(), 'registerUser')
    wrapper.setState({newUser: true})
    wrapper.instance().handleLogin()
    expect(registerUserSpy).toHaveBeenCalled()
  })

  it('should dispatch handlePost when loginUser is called', async () => {
    const url = 'https://turing-feedback-api.herokuapp.com/api/v1/users/login'
    const user = { 
                  name: 'Carlos', 
                  emailInput: 'user@gmail.com'}
    const mockOptions = {
      method: 'POST',
      body: JSON.stringify({ 
              email: 'user@gmail.com'
            }),
      headers: {
        'Content-Type': 'application/json'
      }
    }
    wrapper.instance().loginUser(user)
    expect(wrapper.instance().props.handlePost).toHaveBeenCalledWith(url, mockOptions)
  });

  it('should dispatch handlePost when registerUser is called', async () => {
    const url = 'https://turing-feedback-api.herokuapp.com/api/v1/users/register'
    const user = { 
                  name: 'Carlos', 
                  emailInput: 'user@gmail.com'}
    const mockOptions = {
      method: 'POST',
      body: JSON.stringify({}),
      headers: {
        'Content-Type': 'application/json'
      }
    }
    wrapper.instance().registerUser(user)
    expect(wrapper.instance().props.handlePost).toHaveBeenCalledWith(url, mockOptions)
    })

  it('should call setUser when handleUser is called', () => {
    const mockData = { api_key: 'mockApiKey', role: 'Student'}
    wrapper.instance().handleUser(mockData)
    expect(mockSetUser).toHaveBeenCalled()
  })

  it('should call setRole when handleUser is called', () => {
    const mockData = { api_key: 'mockApiKey', role: 'Student'}
    wrapper.instance().handleUser(mockData)
    expect(mockSetRole).toHaveBeenCalled()
  })

  it.skip('should navigate to /student-dashboard if the role is Student', () => {
    
  })

  it.skip('should navigate to /dashboard if the role is NOT student', () => {

  })

  describe('mapDispatchToProps', () => {
    it('should return setUser to dispatch', () => {
      const user = 'mockApiKey'
      const mockDispatch = jest.fn()
      const actionToDispatch = setUser(user)
      const mappedProps = mapDispatchToProps(mockDispatch)

      mappedProps.setUser(user)

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })

    it('should return handlePost to dispatch', () => {
      const mockUrl = 'www.post.com'
      const mockOptions = {
        method: 'POST',
        body: { name: "name",
                email: "email@gmail.com"
              },
        headers: {
          'Content-Type': 'application/json'
        }
      }
      const mockDispatch = jest.fn()
      const handlePost = jest.fn()
      const actionToDispatch = handlePost(mockUrl, mockOptions)
      const mappedProps = mapDispatchToProps(mockDispatch)

      mappedProps.handlePost(mockUrl, mockOptions)

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })

    it('should return setRole to dispatch', () => {
      const mockRole = 'Student'
      const mockDispatch = jest.fn()
      const actionToDispatch = setRole(mockRole)
      const mappedProps = mapDispatchToProps(mockDispatch)

      mappedProps.setRole(mockRole)

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })
  })
})