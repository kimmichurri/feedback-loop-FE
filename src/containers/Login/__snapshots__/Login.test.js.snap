// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should match the snapshot 1`] = `
<div
  className="login-container"
>
  <NavLink
    to="/dashboard"
  >
    <button
      className="login-button"
      onClick={[Function]}
    >
      Login
    </button>
  </NavLink>
</div>
`;
